objects/src/builder/mod.rs

```diff
@@ -0,0 +1,274 @@
+    }
+}
+
+/// Builder for an `NonFungibleAsset`, the builder can be configured and used multipled times.
+pub struct NonFungibleAssetBuilder<T: Rng> {
+    details_builder: NonFungibleAssetDetailsBuilder<T>,
+}
+
+impl<T: Rng> NonFungibleAssetBuilder<T> {
+    pub fn new(faucet_id: AccountId, rng: T) -> Result<Self, AssetError> {
+        let details_builder = NonFungibleAssetDetailsBuilder::new(faucet_id, rng)?;
+        Ok(Self { details_builder })
+    }
+
+    pub fn with_details(
+        &mut self,
+        details: &NonFungibleAssetDetails,
+    ) -> Result<NonFungibleAsset, AssetError> {
+        NonFungibleAsset::new(&details)
+    }
+
+    pub fn build(&mut self) -> Result<NonFungibleAsset, AssetError> {
+        let details = self.details_builder.build()?;
+        NonFungibleAsset::new(&details)
+    }
+}
+
+/// Builder for an `Account`, the builder allows for a fluent API to construct an account. Each
+/// account needs a unique builder.
+pub struct AccountBuilder<T: Rng> {
+    assets: Vec<Asset>,
+    storage_builder: AccountStorageBuider,
+    code: String,
+    nonce: Felt,
+    account_id_builder: AccountIdBuilder<T>,
+}
+
+const DEFAULT_ACCOUNT_CODE: &str = "\
+use.miden::sat::account
+
+export.incr_nonce
+    push.0 swap
+    # => [value, 0]
+
+    exec.account::incr_nonce
+    # => [0]
+end
+
+export.set_item
+    exec.account::set_item
+    # => [R', V, 0, 0, 0]
+
+    movup.8 drop movup.8 drop movup.8 drop
+    # => [R', V]
+end
+
+export.set_code
+    padw swapw
+    # => [CODE_ROOT, 0, 0, 0, 0]
+
+    exec.account::set_code
+    # => [0, 0, 0, 0]
+end
+
+export.account_procedure_1
+    push.1.2
+    add
+end
+
+export.account_procedure_2
+    push.2.1
+    sub
+end
+";
+
+impl<T: Rng> AccountBuilder<T> {
+    pub fn new(rng: T) -> Self {
+        Self {
+            assets: vec![],
+            storage_builder: AccountStorageBuider::new(),
+            code: DEFAULT_ACCOUNT_CODE.to_string(),
+            nonce: ZERO,
+            account_id_builder: AccountIdBuilder::new(rng),
+        }
+    }
+
+    pub fn add_asset(mut self, asset: Asset) -> Self {
+        self.assets.push(asset);
+        self
+    }
+
+    pub fn add_storage_item(mut self, item: StorageItem) -> Self {
+        self.storage_builder.add_item(item);
+        self
+    }
+
+    pub fn code<C: AsRef<str>>(mut self, code: C) -> Self {
+        self.code = code.as_ref().to_string();
+        self
+    }
+
+    pub fn nonce(mut self, nonce: Felt) -> Self {
+        self.nonce = nonce;
+        self
+    }
+
+    pub fn account_type(&mut self, account_type: AccountType) -> &mut Self {
+        self.account_id_builder.account_type(account_type);
+        self
+    }
+
+    pub fn on_chain(&mut self, on_chain: bool) -> &mut Self {
+        self.account_id_builder.on_chain(on_chain);
+        self
+    }
+
+    pub fn code_root(&mut self, code_root: Digest) -> &mut Self {
+        self.account_id_builder.code_root(code_root);
+        self
+    }
```

Comment:
Similar comment to the one below regarding the `storage_root` setter.  This could result in the accounts code (set via the `pub fn code<C: AsRef<str>>(mut self, code: C)` method) and the accounts code root as perceived by the `AccountIdBuilder` becoming out of sync.

Reply:
Humm, this one seems to have an issue, the `AccountCode::root` depends on the procedure_tree, that in turn depends on the module name, which needs the `AccountId`.

So it seems to create the initial account code we need the id, but to have the id we need the code.

Reply:
Humm, okay, it seems the name is only used to resolve name conflicts. And the results are the MAST roots which should be independent of that. So I will just use a placeholder.

Reply:
That's correct we use the account id to define the namespace for the account module in the assembler.  You're right, it does impact the ergonomics for the creation of new accounts - providing a placeholder is a reasonable solution for now.